src/mo_values/numerics.ml

```diff
@@ -27,6 +27,43 @@ let group_num s =
   Buffer.add_substring buf s exp (len - exp);
   Buffer.contents buf
 
+(* OCaml version of LibTomMath's mp_set_double *)
+let bigint_of_double (f : Wasm.F64.t) : Big_int.big_int =
+  let bits = Wasm.F64.to_bits f in
+
+  (* 0b111...1: 11 bits *)
+  let mask_11_bits = Int64.of_int 0x7FF in
+
+  (* Exponent part of IEEE 754 double, 11 bits *)
+  let exp = Int64.(logand (shift_right_logical bits 52) mask_11_bits) in
+
+  (* Fraction part of IEEE 754 double, 52 bits, but this also sets 53rd bit for
+     some reason. tommath code:
+
+     frac = (cast.bits & (((uint64_t)1 << 52) - (uint64_t)1)) | ((uint64_t)1 << 52); *)
+  let frac = Int64.(logor (shift_right_logical (shift_left bits 12) 12) (shift_left (of_int 1) 52)) in
+
+  if Int64.(equal exp mask_11_bits) then
+    (* Exponent is fully set: NaN or inf *)
+    raise (Invalid_argument "bigint_of_double: argument is NaN or inf");
+
+  (* I don't understand this part, copied from tommath mp_set_double
+     exp -= 1023 + 52; *)
```

Comment:
@ggreif I have an updated version of this function in the thread with all this stuff explained. Could you update this code please? (removed the "automerge" label)

Reply:
Oh, sorry I messed this up. I wanted to copy your code but ended up with the wrong one :-( Thanks for double checking!

I'll fix it.